executor/pipelined_window.go

```diff
@@ -0,0 +1,458 @@
+			if err != nil {
+				return err
+			}
+			e.data[e.dataIdx].remaining -= produced
+			if e.data[e.dataIdx].remaining == 0 {
+				e.dataIdx++
+			}
+		}
+	}
+	if len(e.data) > 0 {
+		chk.SwapColumns(e.data[0].chk)
+		e.data = e.data[1:]
+		e.dataIdx--
+	}
+	return nil
+}
+
+func (e *PipelinedWindowExec) getRowsInPartition(ctx context.Context) (err error) {
+	e.newPartition = true
+	if len(e.rows) == 0 {
+		// if getRowsInPartition is called for the first time, we ignore it as a new partition
+		e.newPartition = false
+	}
+	e.consumed = false
+
+	if e.groupChecker.isExhausted() {
+		var drained, samePartition bool
+		drained, err = e.fetchChild(ctx)
+		if err != nil {
+			err = errors.Trace(err)
+		}
+		// we return immediately to use a combination of true newPartition but 0 in e.rowToConsume to indicate the data source is drained,
+		if drained {
+			e.done = true
+			return nil
+		}
+		samePartition, err = e.groupChecker.splitIntoGroups(e.childResult)
+		if samePartition {
+			// the only case that when getRowsInPartition gets called, it is not a new partition.
+			e.newPartition = false
+		}
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	begin, end := e.groupChecker.getNextGroup()
+	e.rowToConsume += uint64(end - begin)
+	for i := begin; i < end; i++ {
+		e.rows = append(e.rows, e.childResult.GetRow(i))
+	}
+	return
+}
+
+func (e *PipelinedWindowExec) fetchChild(ctx context.Context) (EOF bool, err error) {
+	// TODO: reuse chunks
+	childResult := newFirstChunk(e.children[0])
+	err = Next(ctx, e.children[0], childResult)
+	if err != nil {
+		return false, errors.Trace(err)
+	}
+	// No more data.
+	numRows := childResult.NumRows()
+	if numRows == 0 {
+		return true, nil
+	}
+
+	// TODO: reuse chunks
+	resultChk := chunk.New(e.retFieldTypes, 0, numRows)
+	err = e.copyChk(childResult, resultChk)
+	if err != nil {
+		return false, err
+	}
+	e.accumulated += uint64(numRows)
+	e.data = append(e.data, dataInfo{chk: resultChk, remaining: uint64(numRows), accumulated: e.accumulated})
+
+	e.childResult = childResult
+	return false, nil
+}
+
+func (e *PipelinedWindowExec) copyChk(src, dst *chunk.Chunk) error {
+	columns := e.Schema().Columns[:len(e.Schema().Columns)-e.numWindowFuncs]
+	for i, col := range columns {
+		if err := dst.MakeRefTo(i, src, col.Index); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (e *PipelinedWindowExec) getRow(i uint64) chunk.Row {
+	return e.rows[i-e.rowStart]
+}
+
+func (e *PipelinedWindowExec) getRows(start, end uint64) []chunk.Row {
+	return e.rows[start-e.rowStart : end-e.rowStart]
+}
+
+// finish is called upon a whole partition is consumed
+func (e *PipelinedWindowExec) finish() {
+	e.whole = true
+}
+
+func (e *PipelinedWindowExec) getStart(ctx sessionctx.Context) (uint64, error) {
+	if e.start.UnBounded {
+		return 0, nil
+	}
+	if e.isRangeFrame {
+		var start uint64
+		for start = e.lastStartRow; start < e.rowCnt; start++ {
```

Comment:
Should we save the value e.lastStartRow/e.lastEndRow to avoid calculating the result again and again when the data is not enough?

Reply:
it will be calculated at most every full chunk, i.e. 1024 row, so I think the cost will be averaged and it is okay.